### ISSUE TITLE
Shaded Jersey, Jackson and other libs which expose services can cause conflicts with in-project non-shaded counterparts

### ISSUE THREAD

**Issue by krystiannowak** [testcontainers/testcontainers-java#68]
Due to the fact that the original entries in _META-INF/services_ are still registered to non-shaded names/keys e.g. `javax.ws.rs.ext.MessageBodyWriter` (even though their content is already shaded)
they might be found by auto-discovery/service provider features in corresponding libs and loaded instead of their proper (non-TestContainers) counterparts.

One symptom of such behaviour is the exception given below while integrating Dropwizard with TestContainers library for testing and at the same time using `io.dropwizard.testing.junit.ResourceTestRule` which uses both Jersey and Jackson:
```
org.glassfish.hk2.api.MultiException:
A MultiException has 2 exceptions.  They are:
1. java.lang.ClassCastException: org.testcontainers.shaded.com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider cannot be cast to javax.ws.rs.ext.MessageBodyReader
2. java.lang.IllegalStateException: Unable to perform operation: create on org.glassfish.jersey.message.internal.MessageBodyFactory

Caused by: java.lang.ClassCastException: org.testcontainers.shaded.com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider cannot be cast to javax.ws.rs.ext.MessageBodyReader

WARN  [2016-02-17 13:12:42,266] org.glassfish.jersey.internal.Errors: The following warnings have been detected: WARNING: Unknown HK2 failure detected:
MultiException stack 1 of 2
java.lang.ClassCastException: org.testcontainers.shaded.com.fasterxml.jackson.jaxrs.json.JacksonJsonProvider cannot be cast to javax.ws.rs.ext.MessageBodyReader
        at org.glassfish.jersey.message.internal.MessageBodyFactory.addReaders(MessageBodyFactory.java:507)
        at org.glassfish.jersey.message.internal.MessageBodyFactory.<init>(MessageBodyFactory.java:226)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.glassfish.hk2.utilities.reflection.ReflectionHelper.makeMe(ReflectionHelper.java:1350)
        at org.jvnet.hk2.internal.ClazzCreator.createMe(ClazzCreator.java:271)
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:365)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:471)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:83)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:71)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:97)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:154)
        at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:199)
        at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:122)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2072)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:767)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:706)
        at org.glassfish.jersey.server.model.ComponentModelValidator.<init>(ComponentModelValidator.java:97)
        at org.glassfish.jersey.server.ApplicationHandler.initialize(ApplicationHandler.java:549)
        at org.glassfish.jersey.server.ApplicationHandler.access$500(ApplicationHandler.java:184)
        at org.glassfish.jersey.server.ApplicationHandler$3.call(ApplicationHandler.java:350)
        at org.glassfish.jersey.server.ApplicationHandler$3.call(ApplicationHandler.java:347)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.processWithException(Errors.java:255)
        at org.glassfish.jersey.server.ApplicationHandler.<init>(ApplicationHandler.java:347)
        at org.glassfish.jersey.server.ApplicationHandler.<init>(ApplicationHandler.java:299)
        at org.glassfish.jersey.test.inmemory.InMemoryTestContainerFactory$InMemoryTestContainer.<init>(InMemoryTestContainerFactory.java:77)
        at org.glassfish.jersey.test.inmemory.InMemoryTestContainerFactory$InMemoryTestContainer.<init>(InMemoryTestContainerFactory.java:63)
        at org.glassfish.jersey.test.inmemory.InMemoryTestContainerFactory.create(InMemoryTestContainerFactory.java:111)
        at org.glassfish.jersey.test.JerseyTest.createTestContainer(JerseyTest.java:277)
        at org.glassfish.jersey.test.JerseyTest.setUp(JerseyTest.java:609)
        at io.dropwizard.testing.junit.ResourceTestRule$1.evaluate(ResourceTestRule.java:201)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
MultiException stack 2 of 2
java.lang.IllegalStateException: Unable to perform operation: create on org.glassfish.jersey.message.internal.MessageBodyFactory
        at org.jvnet.hk2.internal.ClazzCreator.create(ClazzCreator.java:392)
        at org.jvnet.hk2.internal.SystemDescriptor.create(SystemDescriptor.java:471)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:83)
        at org.jvnet.hk2.internal.SingletonContext$1.compute(SingletonContext.java:71)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture$1.call(Cache.java:97)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at org.glassfish.hk2.utilities.cache.Cache$OriginThreadAwareFuture.run(Cache.java:154)
        at org.glassfish.hk2.utilities.cache.Cache.compute(Cache.java:199)
        at org.jvnet.hk2.internal.SingletonContext.findOrCreate(SingletonContext.java:122)
        at org.jvnet.hk2.internal.Utilities.createService(Utilities.java:2072)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:767)
        at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:706)
        at org.glassfish.jersey.server.model.ComponentModelValidator.<init>(ComponentModelValidator.java:97)
        at org.glassfish.jersey.server.ApplicationHandler.initialize(ApplicationHandler.java:549)
        at org.glassfish.jersey.server.ApplicationHandler.access$500(ApplicationHandler.java:184)
        at org.glassfish.jersey.server.ApplicationHandler$3.call(ApplicationHandler.java:350)
        at org.glassfish.jersey.server.ApplicationHandler$3.call(ApplicationHandler.java:347)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.processWithException(Errors.java:255)
        at org.glassfish.jersey.server.ApplicationHandler.<init>(ApplicationHandler.java:347)
        at org.glassfish.jersey.server.ApplicationHandler.<init>(ApplicationHandler.java:299)
        at org.glassfish.jersey.test.inmemory.InMemoryTestContainerFactory$InMemoryTestContainer.<init>(InMemoryTestContainerFactory.java:77)
        at org.glassfish.jersey.test.inmemory.InMemoryTestContainerFactory$InMemoryTestContainer.<init>(InMemoryTestContainerFactory.java:63)
        at org.glassfish.jersey.test.inmemory.InMemoryTestContainerFactory.create(InMemoryTestContainerFactory.java:111)
        at org.glassfish.jersey.test.JerseyTest.createTestContainer(JerseyTest.java:277)
        at org.glassfish.jersey.test.JerseyTest.setUp(JerseyTest.java:609)
        at io.dropwizard.testing.junit.ResourceTestRule$1.evaluate(ResourceTestRule.java:201)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:367)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:274)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:161)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:290)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:242)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:121)
```
Due to the fact that apart from already shaded service keys also the non-shaded service keys are still there on the classpath from TestContainers and they (all) are pointing to shaded classes,
they are obviously not the ones which the original library from the project expects to find on the classpath - not from the same implementation but both from non-shaded (Dropwizard's) and
shaded (TestContainers') implementations.

There is a connection between this issue and the issue #35 and it's current solution in ver. 1.0.0 - just to give the context.

One solution would be to remove these non-shaded keys while building shaded TestContainers jar and leave only the shaded keys.
As those shaded keys are added manually (for their value/content only to be shaded during the plugin processing) it seems it's enough to filter out the original keys:
```
diff --git a/core/pom.xml b/core/pom.xml
index 22371f4..f8e4b71 100644
--- a/core/pom.xml
+++ b/core/pom.xml
@@ -129,6 +129,11 @@
                                 <exclude>META-INF/*.SF</exclude>
                                 <exclude>META-INF/*.DSA</exclude>
                                 <exclude>META-INF/*.RSA</exclude>
+                                <exclude>META-INF/services/com.fasterxml.jackson.core.*</exclude>                               
+                                <exclude>META-INF/services/com.github.dockerjava.api.command.*</exclude>
+                                <exclude>META-INF/services/javax.ws.rs.ext.*</exclude>
+                                <exclude>META-INF/services/org.glassfish.hk2.extension.*</exclude>
+                                <exclude>META-INF/services/org.jvnet.hk2.external.generator.*</exclude>
                             </excludes>
                         </filter>
                     </filters>
```

**Comment by rnorth** [comment]
Thanks for this - I thought I'd covered the bases with the shade-tests, but clearly these didn't exercise things enough. Shading is hard :(

I'll accept your PR for merge and release this week; I'd like to add a dropwizard project with resource tests to the shade-test project first (to exhibit the bug and then prove resolution). I can try and work on that tomorrow evening, probably.

Cheers for your hard work on this
Richard

**Comment by rnorth** [comment]
I've just released v1.0.1 which includes your PR. Thanks so much for this - I hope this resolves the issue.
Richard

**Comment by krystiannowak** [comment]
Thanks! It works perfectly now! And it was a pleasure - also to give to, not only to take from OSS.

Cheers,
Krystian
